crates/polars-stream/src/nodes/io_sources/ipc.rs

```diff
@@ -0,0 +1,436 @@
+                    todo!()
+                };
+
+                let source = source.to_memslice()?;
+                read_file_metadata(&mut std::io::Cursor::new(&*source))?
+            },
+        };
+
+        let projection = with_columns
+            .as_ref()
+            .map(|cols| columns_to_projection(cols, &first_metadata.schema))
+            .transpose()?;
+
+        Ok(IpcSourceNode {
+            sources,
+
+            config: IpcSourceNodeConfig {
+                row_index,
+                projection,
+                slice,
+
+                rechunk,
+                include_file_paths,
+
+                first_metadata,
+            },
+
+            is_finished: false,
+        })
+    }
+}
+
+/// Move `slice` forward by `n` and return the slice until then.
+fn slice_take(slice: &mut Range<usize>, n: usize) -> Range<usize> {
+    let offset = slice.start;
+    let length = slice.len();
+
+    assert!(offset < n);
+
+    let chunk_length = (n - offset).min(length);
+    let rng = offset..offset + chunk_length;
+    *slice = 0..length - chunk_length;
+
+    rng
+}
+
+impl ComputeNode for IpcSourceNode {
+    fn name(&self) -> &str {
+        "ipc_source"
+    }
+
+    fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) -> PolarsResult<()> {
+        assert!(recv.is_empty());
+        assert_eq!(send.len(), 1);
+
+        if self.is_finished {
+            send[0] = PortState::Done;
+        }
+
+        if send[0] != PortState::Done {
+            send[0] = PortState::Ready;
+        }
+
+        Ok(())
+    }
+
+    fn spawn<'env, 's>(
+        &'env mut self,
+        scope: &'s TaskScope<'s, 'env>,
+        recv: &mut [Option<RecvPort<'_>>],
+        send: &mut [Option<SendPort<'_>>],
+        _state: &'s ExecutionState,
+        join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+    ) {
+        assert!(recv.is_empty());
+        assert_eq!(send.len(), 1);
+
+        struct BatchMessage {
+            source_idx: usize,
+            memslice: Arc<MemSlice>,
+            metadata: Arc<FileMetadata>,
+            file_path: Option<Arc<str>>,
+            row_idx_offset: IdxSize,
+            slice: Range<usize>,
+            block_range: Range<usize>,
+            seq: u64,
+        }
+
+        let senders = send[0].take().unwrap().parallel();
+        let num_senders = senders.len();
+        let (mut distribute, rxs) =
+            distributor_channel::<BatchMessage>(num_senders, DEFAULT_DISTRIBUTOR_BUFFER_SIZE);
+
+        let config = &self.config;
+        let sources = &self.sources;
+        let is_finished = &mut self.is_finished;
+
+        let decoder_tasks = senders
+            .into_iter()
+            .zip(rxs)
+            .map(|(mut send, mut rx)| {
+                scope.spawn_task(TaskPriority::Low, async move {
+                    let wait_group = WaitGroup::default();
+                    let source_token = SourceToken::new();
```

Comment:
Now that I think about it, it's probably best if all tasks share the same `SourceToken`. I'd suggest making it outside of the loop and cloning it for each sender (it already internally contains an `Arc` which makes clones all share the same stop-requested-state).

Reply:
I'll do the same for the other sources.